cli/src/commands/gerrit.rs

```diff
@@ -0,0 +1,137 @@
+#[derive(Subcommand, Clone, Debug)]
+pub enum GerritCommand {
+    /// "Mail in" a revision to Gerrit; this is the primary way to create
+    /// changes or update changes for code review.
+    ///
+    /// This command modifies each commit in the revset to include a `Change-Id`
+    /// footer in its commit message if one does not already exist. This ID is
+    /// NOT compatible with jj IDs, and is a Gerrit-specific ID.
+    ///
+    /// Note: this command takes a revset as an argument, and will mail in all
+    /// revisions in the revset to Gerrit appropriately, so you may post trees
+    /// or ranges of commits to Gerrit for review.
+    ///
+    /// Note: this command does not actually send an email or use a Mail
+    /// Transfer Agent, but rather uploads the revision to Gerrit for review
+    /// through underlying Git logic. The term "mail" is used to disambiguate
+    /// from the word "submit", which is what Gerrit calls "merging an approved
+    /// change into a target branch", and is used to refer to the concept of
+    /// "mailing in" something to be read by someone else.
+    Mail(MailCommand),
+}
+
+pub fn cmd_gerrit(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    subcommand: &GerritCommand,
+) -> Result<(), CommandError> {
+    match subcommand {
+        GerritCommand::Mail(review) => {
+            cmd_mail(ui, command, review)?;
+        }
+    }
+    Ok(())
+}
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct MailCommand {
+    /// The revset, selecting which commits are mailed in to Gerrit. This can be
+    /// any arbitrary set of commits; they will be modified to include a
+    /// `Change-Id` footer if one does not already exist, and then mailed off to
+    /// Gerrit for review.
+    #[arg(long, short = 'r')]
+    revision: Vec<RevisionArg>,
+
+    /// The origin where your changes will be sent. This MUST be a reference of
+    /// the form `branch@remote`, where `branch` is the target branch name that
+    /// the change will be submitted to, and `remote` is the remote to push to.
+    /// This remote MUST be a Git push URL for your Gerrit instance.
+    #[arg(long = "for", short = 'f', default_value = "main@origin")]
+    for_: String,
+}
+
+pub fn cmd_mail(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    mail: &MailCommand,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let to_mail: IndexSet<Commit> =
+        resolve_multiple_nonempty_revsets(&mail.revision, &workspace_command, ui)?;
+    if to_mail
+        .iter()
+        .any(|commit| commit.id() == workspace_command.repo().store().root_commit_id())
+    {
+        return Err(user_error("Cannot mail the root commit"));
+    }
+
+    workspace_command.check_rewritable(to_mail.iter())?;
+
+    // make sure that the --for parameter is a valid refspec of the
+    // form 'branch@remote'
+    let for_parts: Vec<&str> = mail.for_.split('@').collect();
+    if for_parts.len() != 2 {
```

Comment:
I suppose we could use `RemoteBranchNamePattern` (and its `.is_exact()`). I don't know if that would be simpler or better in some way. You decide.

Reply:
Looks good to me, but I had to shuffle it around. I'll file another PR for moving it out of `branch.rs`

Reply:
#2857